pkg/cluster/service.go

```diff
@@ -0,0 +1,154 @@
+	// create the restclient of fedCluster
+	clusterConfig, err := s.buildClusterConfig(fedCluster, fedCluster.Namespace)
+	if err != nil {
+		return errors.Wrap(err, "cannot create KubeFedCluster config")
+	}
+	cl, err := client.New(clusterConfig, client.Options{})
+	if err != nil {
+		return errors.Wrap(err, "cannot create KubeFedCluster client")
+	}
+
+	cluster := &FedCluster{
+		Name:              fedCluster.Name,
+		Client:            cl,
+		ClusterStatus:     &fedCluster.Status,
+		Type:              Type(fedCluster.Labels[labelType]),
+		OperatorNamespace: fedCluster.Labels[labelNamespace],
+	}
+	if cluster.Type == "" {
+		cluster.Type = Member
+	}
+	if cluster.OperatorNamespace == "" {
+		if cluster.Type == Host {
+			cluster.OperatorNamespace = defaultHostOperatorNamespace
+		} else {
+			cluster.OperatorNamespace = defaultMemberOperatorNamespace
+		}
+	}
+
+	clusterCache.addFedCluster(cluster)
+	return nil
+}
+
+// DeleteKubeFedCluster takes the KubeFedCluster CR object
+// and deletes FedCluster instance that has same name from a cache (if exists)
+func (s *KubeFedClusterService) DeleteKubeFedCluster(obj interface{}) {
+	cluster, ok := obj.(*v1beta1.KubeFedCluster)
+	if !ok {
+		err := fmt.Errorf("incorrect type of KubeFedCluster: %+v", obj)
+		s.Log.Error(err, "cluster not deleted")
+		return
+	}
+	log := s.enrichLogger(cluster)
+	log.Info("observed a deleted cluster")
+	clusterCache.deleteFedCluster(cluster.Name)
+}
+
+// UpdateKubeFedCluster takes the KubeFedCluster CR object,
+// creates FedCluster with a kube client and stores it in a cache.
+// If there cache already contains such an instance, then it is overridden.
+func (s *KubeFedClusterService) UpdateKubeFedCluster(_, newObj interface{}) {
+	newCluster, ok := newObj.(*v1beta1.KubeFedCluster)
+	if !ok {
+		err := fmt.Errorf("incorrect type of new KubeFedCluster: %+v", newObj)
+		s.Log.Error(err, "cluster not updated")
+	}
+	log := s.enrichLogger(newCluster)
+	log.Info("observed an updated cluster")
+
+	err := s.addKubeFedCluster(newCluster)
+	if err != nil {
+		log.Error(err, "the cluster was not updated")
+	}
+}
+
+func (s *KubeFedClusterService) enrichLogger(cluster *v1beta1.KubeFedCluster) logr.Logger {
+	return s.Log.
+		WithValues("Request.Namespace", cluster.Namespace, "Request.Name", cluster.Name)
+}
+
+func (s *KubeFedClusterService) buildClusterConfig(fedCluster *v1beta1.KubeFedCluster, fedNamespace string) (*rest.Config, error) {
+	clusterName := fedCluster.Name
+
+	apiEndpoint := fedCluster.Spec.APIEndpoint
+	if apiEndpoint == "" {
+		return nil, errors.Errorf("The api endpoint of cluster %s is empty", clusterName)
+	}
+
+	secretName := fedCluster.Spec.SecretRef.Name
+	if secretName == "" {
+		return nil, errors.Errorf("Cluster %s does not have a secret name", clusterName)
+	}
+	secret := &v1.Secret{}
+	name := types.NamespacedName{Namespace: fedNamespace, Name: secretName}
+	err := s.Client.Get(context.TODO(), name, secret)
+	if err != nil {
+		return nil, err
```

Comment:
do you need to wrap the error with a contextual message?

Reply:
done 717dcc1